src/framework/components/rigid-body/component.js

```diff
@@ -646,8 +646,11 @@ Object.assign(pc, function () {
 
                     var p = ammoTransform.getOrigin();
                     var q = ammoTransform.getRotation();
-                    this.entity.setPosition(p.x(), p.y(), p.z());
                     this.entity.setRotation(q.x(), q.y(), q.z(), q.w());
+                    var rot = this.entity.getRotation();
+
+                    var rotatedOffset = rot.transformVector(this.entity.collision.offset);
+                    this.entity.setPosition(p.x() - rotatedOffset.x, p.y() - rotatedOffset.y, p.z() - rotatedOffset.z);
```

Comment:
This is a performance critical part of the engine. I can think of a couple of optimizations here. For example, you only have to do the additional math if the offset is non-zero. Also, since Ammo is provided the rotation, you don't need to get it from PlayCanvas' `getRotation` function. You can just do `ammoQuat.set(p.x(), p.y(), p.z(), p.w());`. Also, since you're calling `transformVector` without a result parameter, one will be allocated internally which will generate lots of garbage. So again, use a temporary declared at the top of the file: `ammoVec1`.